### PR TITLE
[no squash] Handle tile generation edge cases

### PR DESCRIPTION
**Pull request by sfan5**
addresses #14306

## To do

This PR is Ready for Review.

## How to test

```lua
core.after(1, function()
	local s = "[combine:16x16:0,0="
	core.show_formspec("singleplayer", "formimage:image",
		"size[4,5]image[0,0;4,8;" .. core.formspec_escape(s) .. "]"
	)
end)
```

you should get an error or two

### CODE REVIEW

File: src/client/tile.cpp [hunk 1]
```diff
@@ -1834,15 +1865,12 @@ bool TextureSource::generateImagePart(std::string part_of_name,
 			to produce a valid string.
 		*/
 		else if (str_starts_with(part_of_name, "[png:")) {
-			Strfnd sf(part_of_name);
-			sf.next(":");
 			std::string png;
 			{
-				std::string blob = sf.next("");
+				std::string blob = part_of_name.substr(5);
```
Comment (appgurueu) [inline src/client/tile.cpp:1870]: BTW we can make this use a string view with a trivial modification to our base64 util.

**Review by appgurueu — APPROVED** [review]
Looks fine. Some thoughts:

Regarding the first commit (deduplication):

* Dunno about const ref -> const copy for stuff like colors (in general). It shouldn't make a difference for `static` methods performance-wise anyways, because the C++ compiler can just drop the refs here (see e.g. [on GodBolt](https://godbolt.org/z/j8raEKW64)). Anyways, I don't care either way, both are fine.
* The macros are okay. Is the `undef` for limiting their scope necessary? That just prevents that code further down in the source file makes use of them. (BTW, I think a better solution might be to use exceptions. Then you could use `throw`ing functions inside of macros with a `return false`.)

Regarding the second commit:

The checks look fine to me. I get two acceptable errors in chat for the texture modifier.

(BTW, there's a bug in `lua_api.md` regarding the docs of `[fill`: "Creates a texture of the given size and color, optionally with an , position"; you could fix that here, or I could fix it in a "misc. stuff" PR I may open soon).

Third commit is obviously fine.

File: src/client/tile.cpp (hunk 1 above)
Comment (sfan5) [reply src/client/tile.cpp:1870]: I'm aware. There's a bunch more code that would immediately benefit from string_view.
I might have time to prepare a PR in a 1-2 weeks, but feel free to also do it.

**Comment by sfan5** [comment]
> * Dunno about const ref -> const copy for stuff like colors (in general).

`SColor` is a single 32-bit number. Trying to pass it by pointer is always strictly worse.

> (BTW, I think a better solution might be to use exceptions. 

I guess, but I wasn't planning to rework this code entirely. There's much more that could be improved.

File: src/client/tile.cpp [hunk 2]
```diff
@@ -1075,6 +1062,12 @@ video::IImage* TextureSource::generateImage(const std::string &name, std::set<st
 	if (baseimg == NULL) {
 		errorstream << "generateImage(): baseimg is NULL (attempted to"
 				" create texture \"" << name << "\")" << std::endl;
+	} else if (baseimg->getDimension().Width == 0 ||
+		baseimg->getDimension().Height == 0) {
```
Comment (SmallJoker) [inline src/client/tile.cpp:1066]:
```suggestion
	} else if (baseimg->getDimension().getArea() == 0) {
```

File: src/client/tile.cpp [hunk 3]
```diff
@@ -1183,55 +1176,76 @@ void blitBaseImage(video::IImage* &src, video::IImage* &dst)
...
+
 bool TextureSource::generateImagePart(std::string part_of_name,
 		video::IImage *& baseimg, std::set<std::string> &source_image_names)
 {
 	const char escape = '\\'; // same as in generateImage()
 	video::IVideoDriver *driver = RenderingEngine::get_video_driver();
 	sanity_check(driver);
 
+	if (baseimg && (baseimg->getDimension().Width == 0 ||
+		baseimg->getDimension().Height == 0)) {
```
Comment (SmallJoker) [inline src/client/tile.cpp:1212]:
```suggestion
	if (baseimg && baseimg->getDimension().getArea() == 0) {
```
also "fixes" missing indent level

File: src/client/tile.cpp [hunk 4]
```diff
@@ -1661,8 +1629,8 @@ bool TextureSource::generateImagePart(std::string part_of_name,
 						img->getDimension());
 				img->drop();
 			} else {
-				errorstream << "generateImage(): Failed to load \""
-						<< filename << "\".";
+				errorstream << "generateImagePart(): Failed to load image \""
+						<< filename << "\" for [mask";
```
Comment (SmallJoker) [inline src/client/tile.cpp:1633]:
```suggestion
						<< filename << "\" for [mask" << std::endl;
```

File: src/client/tile.cpp [hunk 5]
```diff
@@ -2057,8 +1967,8 @@ bool TextureSource::generateImagePart(std::string part_of_name,
 						img->getDimension(), hardlight);
 				img->drop();
 			} else {
-				errorstream << "generateImage(): Failed to load \""
-					<< filename << "\".";
+				errorstream << "generateImage(): Failed to load image \""
+					<< filename << "\" for [overlay or [hardlight";
```
Comment (SmallJoker) [inline src/client/tile.cpp:1971]:
```suggestion
					<< filename << "\" for [overlay or [hardlight" << std::endl;
```

**Review by SmallJoker — COMMENTED** [review]
LGTM overall
